### PR TITLE
8299234: JMX Repository.query performance

### PR DESCRIPTION
**Pull request by alexeybakhtin**
Please find a patch to improve JMX Repository.query performance

Using ObjectName.apply() allows significantly decrease memory usage and the number of GC cycles:
Before:
```
$ java test 1000000 1000000
Test PASSED in 8943169791 ns.
GC: G1 Young Generation getCollectionCount()=177 getCollectionTime()=118
```

After:
```
$ java test 1000000 1000000
Test PASSED in 4808213917 ns.
GC: G1 Young Generation getCollectionCount()=88 getCollectionTime()=53
```
Private ObjectName.matchDomains() method is also updated to minimize unnecessary memory allocation.

All javax/management jtreg tests passed successfully.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299234](https://bugs.openjdk.org/browse/JDK-8299234): JMX Repository.query performance

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11758/head:pull/11758` \
`$ git checkout pull/11758`

Update a local copy of the PR: \
`$ git checkout pull/11758` \
`$ git pull https://git.openjdk.org/jdk pull/11758/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11758`

View PR using the GUI difftool: \
`$ git pr show -t 11758`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11758.diff">https://git.openjdk.org/jdk/pull/11758.diff</a>

</details>

### CODE REVIEW

**Comment by kevinjwalls** [comment]
Hi,
Speedup sounds good.
So we don't need the ObjectNamePattern class, and it's private so no dependencies.
I haven't fully followed the wildcard matching changes yet.
Will try and do some testing also.

**Comment by dfuch** [comment]
I see several issues with the proposed changes: ObjectName is not final, so I'm not sure the changes proposed to ObjectName are safe when/if a subclass is supplied. And in addition, because it's not final, you will need to make a defensive copy of the input parameter in Repository::query and do the matching against the copy.

**Comment by alexeybakhtin** [comment]
> I see several issues with the proposed changes: ObjectName is not final, so I'm not sure the changes proposed to ObjectName are safe when/if a subclass is supplied. And in addition, because it's not final, you will need to make a defensive copy of the input parameter in Repository::query and do the matching against the copy.

Hi @dfuch, Thank you for review!

File: src/java.management/share/classes/javax/management/ObjectName.java
Comment (dfuch) [inline src/java.management/share/classes/javax/management/ObjectName.java:2024]: This looks wrong. Look at the parameters in wildmatch: 
```
wildmatch(final String str, final String pat,
                   int stri, final int strend, int pati, final int patend)
```

What tests did you run to validate your changes? If none of the ObjectName tests fail with your changes you will need to add a new test that catches the error.

**Review by dfuch — COMMENTED** [review]
Using `ObjectName::getInstance` for the defensive copy is the right choice since it will only make a copy if the argument is a subclass. That looks good. I see your point about using `_canonical_name`. Technically this is still an observable change of behaviour - so it might be preferable to file a CSR if you want to proceed with the changes made to `ObjectName`. 
Regardless of that these changes seem to be introducing a bug (see my other comment inline).

File: src/java.management/share/classes/javax/management/ObjectName.java
Comment (alexeybakhtin) [reply src/java.management/share/classes/javax/management/ObjectName.java]: Thank you again for review and finding.
Current tests do not catch such issue, so I extended ApplyWildcardTest.java with domain patterns. The test passed on the original and new code, fails with incorrect Util.wildmatch parameters

**Comment by dfuch** [comment]
Hi Alexey - you have addressed most of my concerns but let me reinstate that the changes to ObjectName are observable by subclasses and thus will require a CSR.

**Comment by alexeybakhtin** [comment]
Hi Daniel, Yes, I'd like to backport these changes to JDK11 and 8. Such changes to the OjectName could be unacceptable risky. I have reverted all changes in the ObjectName class.

**Review by dfuch — COMMENTED** [review]
I don't see an obvious issue with the proposed changes. Do you have a JMH benchmark to prove the new code is faster? Also please obtain a review from a maintainer of the servicieability area before integrating.

**Comment by alexeybakhtin** [comment]
@sspitsyn, Could you take a look at the proposed enhancement or suggest someone who can review it? Thank you

**Comment by alexeybakhtin** [comment]
@AlanBateman, Could you please take a look at the proposed performance improvement? It is required to reduce high CPU utilization caused by unnecessary HashTable cloning in the [ObjectName::getKeyPropertyList()](https://github.com/openjdk/jdk/blob/master/src/java.management/share/classes/javax/management/ObjectName.java#L1685). I also would like to backport it to 17, 11 and 8 if the changes are acceptable.
Thank you in advance

**Comment by sspitsyn** [comment]
Hi Alexey,
I hope, Kevin will have a chance to review your PR. I'll check with him.

**Comment by alexeybakhtin** [comment]
> Hi Alexey, I hope, Kevin will have a chance to review your PR. I'll check with him.

Thank you!

**Comment by kevinjwalls** [comment]
Yes, I think it looks good.
I looked a while ago but it was somewhat unfamiliar so was glad of Daniel's input.

It's a bit of a style nit but our ifs always use brackets, so I miss that in ObjectName.matchDomains(), particularly as the statements are multi-line.

I tested the change with test/jdk/javax/management/ and everything still passes.
The repo is somewhat behind, but these changes have no merge problems with the latest. I took these changes into a latest jdk repo to build and test.

File: src/java.management/share/classes/javax/management/ObjectName.java
Comment (kevinjwalls) [inline src/java.management/share/classes/javax/management/ObjectName.java:2026]: 2026 and 2031 I was referring to with the nit about if and { }

Comment (alexeybakhtin) [reply src/java.management/share/classes/javax/management/ObjectName.java]: > 2026 and 2031 I was referring to with the nit about if and { }

Hi @kevinjwalls ! Thank you for review.
The braces are added

Comment (kevinjwalls) [reply src/java.management/share/classes/javax/management/ObjectName.java]: Thanks!

**Review by kevinjwalls — APPROVED** [review]

**Comment by alexeybakhtin** [comment]
@kevinjwalls, Thank you very much for reviewing.
@sspitsyn, the patch is approved by Kevin but unfortunately, he is not a reviewer in JDK project. Could you please approve it or ask somebody from the serviceability group? Thank you in advance.

**Comment by sspitsyn** [comment]
> @kevinjwalls, Thank you very much for reviewing. @sspitsyn, the patch is approved by Kevin but unfortunately, he is not a reviewer in JDK project. Could you please approve it or ask somebody from the serviceability group? Thank you in advance.

Alexey, two reviews are normally required before integration in the Serviceability area.
I guess, you need to wait for review from Daniel Fuchs (an expert in this area) or someone else.

**Review by dfuch — APPROVED** [review]
This last version LGTM

**Pull request closed without merging**
